python/tests/test_writer.py

```diff
@@ -496,3 +496,218 @@ def test_writer_with_options(tmp_path: pathlib.Path):
+                }
+            ),
+            "expected_case1": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array(
+                        [
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                        ],
+                        pa.date32(),
+                    ),
+                    "val": pa.array([2, 2, 1, 1], pa.int64()),
+                }
+            ),
+            "filter2": [("p2", ">", "2022-01-01")],
+            "data_case2": pa.table(
+                {
+                    "p1": pa.array(["1", "2"], pa.string()),
+                    "p2": pa.array([date(2022, 1, 2), date(2022, 1, 2)], pa.date32()),
+                    "val": pa.array([3, 3], pa.int64()),
+                }
+            ),
+            "expected_case2": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array(
+                        [
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                        ],
+                        pa.date32(),
+                    ),
+                    "val": pa.array([2, 3, 1, 3], pa.int64()),
+                }
+            ),
+        },
+    ],
+)
+def test_partition_overwrite(tmp_path: pathlib.Path, cases_data: Dict[str, pa.Array]):
+    sample_data = cases_data["init"]
+    write_deltalake(
+        str(tmp_path), sample_data, mode="overwrite", partition_by=["p1", "p2"]
+    )
+
+    delta_table = DeltaTable(str(tmp_path))
+    assert (
+        delta_table.to_pyarrow_table().sort_by(
+            [("p1", "ascending"), ("p2", "ascending")]
+        )
+        == sample_data
+    )
+
+    sample_data = cases_data["data_case1"]
+    expected_data = cases_data["expected_case1"]
+    write_deltalake(
+        str(tmp_path),
+        sample_data,
+        mode="overwrite",
+        partitions_filters=cases_data["filter1"],
+    )
+
+    delta_table.update_incremental()
+    assert (
+        delta_table.to_pyarrow_table().sort_by(
+            [("p1", "ascending"), ("p2", "ascending")]
+        )
+        == expected_data
+    )
+
+    sample_data = cases_data["data_case2"]
+    expected_data = cases_data["expected_case2"]
+
+    write_deltalake(
+        str(tmp_path),
+        sample_data,
+        mode="overwrite",
+        partitions_filters=cases_data["filter2"],
+    )
+    delta_table.update_incremental()
+    assert (
+        delta_table.to_pyarrow_table().sort_by(
+            [("p1", "ascending"), ("p2", "ascending")]
+        )
+        == expected_data
+    )
+
+
+def test_partition_overwrite_unfiltered_data_fails(tmp_path: pathlib.Path):
```

Comment:
Could we add a test that validates we get an error if you pass a partition filter that contains columns not in the partition columns?

Reply:
Also what happens if you pass them to a non-partitioned table?

Reply:
Added test cases